WordPress/gutenberg
# Patterns: merge unsynced into inserter patterns tab and add paging and filtering

## What?
Merges the synced patterns tab into the main patterns tab and adds an `All patterns` category and paging and filtering in the patterns inserter tab and patterns explorer.

## Why?
Currently it is confusing that synced patterns are in a separate tab, and also causes performance issues because patterns are not paged.

## How?

- Removes the synced patterns tab from inserter
- Updates the store patterns selector to get synced and unsynced patterns so both display in the existing patterns tab
- Adds paging to the patterns list to account for large numbers of synced patterns in some sites
- Adds a filter to allow users to limit pattern list to just theme, or user patterns, and another filter when user patterns selected to limit to synced or unsynced patterns

This PR was part of a group of PRs that were merged back into https://github.com/WordPress/gutenberg/pull/53835 before being merged into trunk. This PR was branched off https://github.com/WordPress/gutenberg/pull/53933.

A full list of the PRs involved in this work can be [found here](https://github.com/WordPress/gutenberg/issues/53164).

## Testing Instructions

- In the post editor add some synced and unsynced patterns with categories assigned
- Check that the patterns appear in the correct categories in the inserter patterns tab
- Check that the `All patterns` tab shows all patterns and that paging at the bottom of the tab works
- Check that the Source Filter select list works as expected and that correct patterns display for each selected filter
- Also check that when the `My patterns` source filter is selected that a `Sync type filter` appears at the top of the patterns list and works as expected
- Check that changing Source or Sync type filters when categories are selected resets the category to `All`
- Repeat all of the above in the patterns explorer modal
- With screenreader/voiceover running try selecting a pattern source from dropdown at top of patterns tab and check that the number of category buttons displayed is announced correctly

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/3629020/e377fe33-675e-4344-bdfd-e63141e74a3f
